Add step-by-step tutorial to making institutional configs (#734)
nf-core/nf-co.re: markdown/usage/tutorials/step_by_step_institutional_profile.md

---
title: Step-by-step guide to writing an institutional profile
subtitle: Walkthrough on what you need to set up an nf-core institutional profile
---

## Introduction

nf-core offers a centralised place to store Nextflow configuration profiles that work at an _institutional_ level.

What this means is that you can specify common Nextflow pipeline configurations and options that can be shared across all users of your particular institutional cluster.

> nf-core offers two level of profile sharing: institutional and pipeline-specific profiles via [nf-core/configs](https://github.com/nf-core/configs)

Reviewer: I would rephrase this part, I would not categorise into institutional and pipeline-specific, but global and pipeline-specific. 

```suggestion
nf-core offers two level of institutional profile sharing: global and pipeline-specific profiles via [nf-core/configs](https://github.com/nf-core/configs)
```

Pipeline-specific are still institutional, (except for the viralrecon genomes.config file that is global but pipeline specific.

Reviewer: But I recognise the fact that we actually use institutional most of the time to mean what I called global

Author: What about: `global institutional config` vs . `pipeline-specific institutional config` throughout the text?

Reviewer: Sounds good to me, maybe we should update the nf-core/configs repo for that too?

Author: Probably should! Will do after this is in

- Institutional configuration represent configuration options that apply to users of _all_ nf-core pipelines. These typically define settings regarding the cluster itself, such as the type of scheduler being used, maximum resource limits and so on.
- Pipeline-specific profiles that represent configuration options that apply to users of a _specific_ nf-core pipelines. These typically define common parameters all users of the pipeline would use, or customise resource requirements for particular processes of that specific pipeline

This walkthrough will guide you through making an _institutional_-level profile. It will:

- Describe commonly useful information that is worth gathering _before_ writing such a profile
- Go through each component of the profile to describe how this should be written
- Show how to test such a profile before submitting it to nf-core

## Preparation

We will first go through a list of questions to ask and what information to gather prior writing the profile. Not all questions will be relevant to your cluster, however it should cover a range of common set of cluster specifications.

> ## Does an institutional profile already exist

Reviewer: ```suggestion
### Does an institutional profile already exist
```

> First you should check nf-core/configs an institutional config exists! If you see any issues with it, or it's not working for you - contact the person who originally made it.

Reviewer: ```suggestion
First you should check nf-core/configs an institutional profile exists! If you see any issues with it, or it's not working for you - contact the person who originally made it.
```

> If no config already exists, continue with this tutorial!

Reviewer: ```suggestion
If no profile already exists, continue with this tutorial!
```

### At what level should the institutional profile be designed

Various institutions have different structures in the way clusters are accessed by different groups.

This is an important consideration to make before starting writing a profile because you need to select a useful and descriptive name.

> Sometimes one 'institution' will have multiple 'clusters' with the different names, which may share a common storage space. In this case it might make sense to have one nf-core configuration file for the whole institution, but with multiple _internal_ profiles representing each cluster.
> 
> Alternatively, within one 'institution', you may have groups or departments with entirely different and separated clusters with dedicated IT teams. In this case it may make sense to have separate configuration files without a common institutional 'umbrella' name.

Reviewer: ```suggestion
Sometimes one institution will have multiple clusters with the different names, which may share a common storage space. In this case it might make sense to have one nf-core configuration file for the whole institution, but with multiple _internal_ profiles representing each cluster.

Alternatively, within one institution, you may have groups or departments with entirely different and separated clusters with dedicated IT teams. In this case it may make sense to have separate configuration files without a common institutional 'umbrella' name.
```

### Does your cluster use a scheduler or executor

Nextflow has integrated a range of scheduling systems to simplify running Nextflow pipelines on a range of HPCs. What this means is that it will write and sumbit submission scripts to your given scheduler or 'executor' (in Nextflow language) for you.

You can see the range of support schedulers on the [Nextflow documentation](https://www.Nextflow.io/docs/latest/executor.html). Check if your scheduler is there, and how it would be written (typically the name in all lower case).

You should note down if there are any special options that your cluster's scheduler requires. This maybe special ways in which memory requirements are specified, e.g.: memory per core, types of parallelisation library is required etc., the maximum number of parallel jobs that are allowed.

<details markdown='1'><summary> SGE tips</summary>

> Some SGE configurations require the use of the `-l h_vmem=` flag _in addition_ to `-l virtual_free=` - the latter of which is the default used by default in Nextflow.

Reviewer: `the default used by default`?
```suggestion
Some SGE configurations require the use of the `-l h_vmem=` flag _in addition_ to `-l virtual_free=` - the latter of which is used by default in Nextflow.
```

```bash
sinfo
```

and look for the TIMELIMIT column for the longest running queue that _all_ users have access to.
</details>
</br>

### Do you have to specify a queue or partition

Some clusters require your to specify which queue or partition to submit your job to (based on the resource requirements of the job). You should check if these exist, what they are called, and what limits they may have (e.g. walltimes).

You should already consider in which 'order' you would like to use these queues, to make use of the Nextflow and nf-core automated 'retry' system (i.e. if a process runs too long or uses too much memory, it can be automatically resubmitted with more time/memory)

<details markdown='1'><summary> SLURM tips</summary>

For example, if you're using the SLURM scheduler, you can run the following command to the walltime of any queues/partitions on your cluster with

```bash
sinfo
```

You should record the names of each `PARTITION` and the corresponding `TIMELIMIT` column. Note there may be additional resource limitations for each queue, so check your institutional cluster's documentation for any other limits.

</details>

> ## Are there any common resource or data locations on the cluster

Reviewer: ```suggestion
### Are there any common resource or data locations on the cluster
```

> For institutions that work on common topics (e.g. genomics), you might have a centralised local copy of the Illumina `iGenome` set of reference genomes. If it exists, you can note down the paths to this directory.

Reviewer: ```suggestion
For institutions that work on common topics (e.g. genomics), you might have a centralised local copy of the AWS `iGenomes` set of reference genomes. If it exists, you can note down the paths to this directory.
```

> Furthermore, some institutions will specify a common place to store all container images (e.g. for Docker, Singularity etc) in a cache directory. This helps preventing multiple users having their own copies of the same pipeline container image. You should find out if, and where this location exists and note down the path.

Reviewer: I don't think there is a cache dir option for Docker

Author: Huh, just checked the nextflow docs and you're right. That's wierd.

> ## Should jobs be run using a local `$TMPDIR`

Reviewer: ```suggestion
### Should jobs be run using a local `$TMPDIR`
```

Some clusters require the use of a temporary _per-node_ `scratch` directory, the location of which is specified in an environmental variable `$TMPDIR`. Check if this is the case and this is set up in `$TMPDIR`

> ## Preparation checklist

Reviewer: ```suggestion
### Preparation checklist
```

> To summarise, before you start writing the profile have you:
> 
> - Checked you know what institutional level you will make the profile at
> - Checked permission to publicly post the profile
> - Checked which scheduler the cluster uses
> - Checked which container engine you will specify
> - Checked which maximum computational resource limits exist
> - Checked for any queue/partition names and specifications
> - Checked for locations of any common resources or cache directories
> - Checked if jobs on the cluster require use of a per-node `scratch`

Reviewer: ```suggestion
To summarise, before you start writing the profile have you checked:

- At what institutional level you will make the profile
- The permission to publicly post the profile
- Which scheduler the cluster uses
- Which container engine you will specify
- Which maximum computational resource limits exist
- Queue/partition names and specifications
- Locations of any common resources or cache directories
- If jobs on the cluster require use of a per-node `scratch`
```

## Creating the profile

> Once you have all the information ready, we can begin writing the institutional profile! The tutorial will now go through each relevant Nextflow 'scope' block and explain what and when to put information in.

Reviewer: ```suggestion
Once you have all the information ready, we can begin writing the institutional profile! The walkthrough will now go through each relevant Nextflow 'scope' block and explain what and when to put information in.
```

Before you start consider what your institutional profile should be called. The names should be as unique and recognisable as possible - however abbreviations or acronyms are fine.

For example, the name of the cluster can be used (e.g. [`phoenix`](https://github.com/nf-core/configs/blob/master/conf/phoenix.config) or [binac](https://github.com/nf-core/configs/blob/master/conf/binac.config)). In some cases an acronym of the cluster or the institute can be selected (e.g. [`SHH`](https://github.com/nf-core/configs/blob/master/conf/shh.config) or [`EVA`](https://github.com/nf-core/configs/blob/master/conf/eva.config)).

> Next, make sure to make your own _fork_ of [nf-core/configs](https://github.com/nf-core/configs), and make a new branch to work on. It's normally best to call the branch the name of your proposed cluster name. Clone this to your machine, and open the folder.

Reviewer: It's not necessary to make a branch, usually people would work on their own fork of nf-core/configs once, and that's all.
In any other case, I'd recommend making a branch, in this case I think it's less important.
But as it's a good practice, I'm all for it.

Author: I also originally did that until recently when I got stuck doing loads of debugging and messed up master, so then couldn't pull in the latest nf-core/configs master. So I think while it's less likely, I feel it's good practise anyway (like you said) 😬

In the `conf/<your_cluster_name>.config` file, add to a params scope something like the following:

```nextflow
params {
  config_profile_description = '<your_cluser_name> cluster profile provided by nf-core/configs.'
  config_profile_contact = '<your_name> (<your_github_handle>)'
  config_profile_url = 'https://<institutional_url>.com'
}
```

Next, in the same scope, we can also specify the `max_*` series of params. These are used by nf-core pipelines to limit automatic resubmission of resource-related failed jobs to ensure submitted retries do not exceed the maximum available on your cluster. These values should be the ones you found for the largest node of your cluster (i.e., the largest node a user's job can be submitted to). For example:

```nextflow
params {
  config_profile_description = '<your_cluser_name> cluster profile provided by nf-core/configs.'
  config_profile_contact = '<your_name> (<your_github_handle>)'
  config_profile_url = 'https://<institutional_url>.com'
  max_memory = 2.TB
  max_cpus = 128
  max_time = 720.h
}
```

> Finally, if you have a common resource directory for the iGenomes collection of reference genomes, this can can also go in the `params{}` scope.

Reviewer: ```suggestion
Finally, if you have a common resource directory for the iGenomes collection of reference genomes, this can can also go in the `params` scope.
```

  executor = 'sge'
  queue = { task.cpus > 24 ? 'big' : 'small' }
  maxRetries = 2
  clusterOptions = { '-l h_vmem=${task.memory.toGiga()}G' }
  beforeScript = 'module load singularity'
}

executor {
  queueSize = 8
}
```

Where a given Nextflow run can only have 8 submitted jobs at once (and will wait to one job is completed before submitting the next one).

A similar directive is the `submitRateLimit` which specifies how many jobs can be specified in a given time frame (as some clusters may penalise you for over-submitting)

```nextflow
params {
  config_profile_description = '<your_cluser_name> cluster profile provided by nf-core/configs.'
  config_profile_contact = '<your_name> (<your_github_handle>)'
  config_profile_url = 'https://<institutional_url>.com'
  max_memory = 2.TB
  max_cpus = 128
  max_time = 720.h
  igenomes_base = '/<path>/<to>/igenomes/'
}

process {
  executor = 'sge'
  queue = { task.cpus > 24 ? 'big' : 'small' }
  maxRetries = 2
  clusterOptions = { '-l h_vmem=${task.memory.toGiga()}G' }
  beforeScript = 'module load singularity'
}

executor {
  queueSize = 8
  submitRateLimit = '10 sec'
}
```

Where a maximum of 10 jobs can be submitted per second

> #### <your_container\> scope

Reviewer: I'll go for singularity/docker/podman scope here directly.
Or maybe just container scope in the title, and explain that it's either singularity, docker, podman, or even conda...

```nextflow
params {
  config_profile_description = '<your_cluser_name> cluster profile provided by nf-core/configs.'
  config_profile_contact = '<your_name> (<your_github_handle>)'
  config_profile_url = 'https://<institutional_url>.com'
  max_memory = 2.TB
  max_cpus = 128
  max_time = 720.h
  igenomes_base = '/<path>/<to>/igenomes/'
}

process {
  executor = 'sge'
  queue = { task.cpus > 24 ? 'big' : 'small' }
  maxRetries = 2
  clusterOptions = { '-l h_vmem=${task.memory.toGiga()}G' }
  beforeScript = 'module load singularity'
}

executor {
  queueSize = 8
  submitRateLimit = '10 sec'
}

singularity {
  enabled = true
  autoMounts = true
  cacheDir = '/<path>/<to>/<your>/<image_cache>'
}
```

Each container engine or software environment may have different options, so be sure to check the Nextflow documentation what options you may have.

> #### profiles{} scope

Reviewer: ```suggestion
#### profiles scope
```
why adding the `{}`?

In some cases, you may want to define multiple contexts that have different specifications - For example, your institution may have has two distinct clusters or set of nodes with slightly different specifications, while still sharing the same storage.

> One concise way of specifying these is via the `profiles{}` scope. These essentially act as nested configs, where you can extend or modify the base parameters set in the main config.

Reviewer: ```suggestion
One concise way of specifying these is via the `profiles` scope. These essentially act as nested configs, where you can extend or modify the base parameters set in the main config.
```

```nextflow
process {
  executor = 'sge'
  queue = { task.cpus > 24 ? 'big' : 'small' }
  maxRetries = 2
  clusterOptions = { '-l h_vmem=${task.memory.toGiga()}G' }
  beforeScript = 'module load singularity'
}

executor {
  queueSize = 8
  submitRateLimit = '10 sec'
}

singularity {
  enabled = true
  autoMounts = true
  cacheDir = '/<path>/<to>/<your>/<image_cache>'
}

profiles {
  red {
    params {
      config_profile_description = '<your_institution_name> 'red' cluster cluster profile provided by nf-core/configs.'
      config_profile_contact = '<your_name> (<your_github_handle>)'
      config_profile_url = 'https://<institutional_url>.com'
      max_memory = 2.TB
      max_cpus = 128
      max_time = 720.h
      igenomes_base = '/<path>/<to>/igenomes/'
    }
  }

  blue {
    params {
      config_profile_description = '<your_institution_name> 'blue' cluster profile provided by nf-core/configs.'
      config_profile_contact = '<your_name> (<your_github_handle>)'
      config_profile_url = 'https://<institutional_url>.com`'
      max_memory = 256.GB
      max_cpus = 64
      max_time = 24.h
      igenomes_base = '/<path>/<to>/igenomes/'
    }
  }
}
```

> You can see here we have moved the `params{}` block into each of the _profiles_, and updated the `config_profile_description` and `max_*` parameters accordingly.

Reviewer: ```suggestion
You can see here we have moved the `params` block into each of the _profiles_, and updated the `config_profile_description` and `max_*` parameters accordingly.
```

## Testing the profile

Once all of your files are prepared, it's time to test your institutional config. All nf-core pipelines come with a special parameter called `--custom_config_base`. This allows you to override the pipeline run from loading the nf-core/configs from the nf-core repository, and you can instead get it to look in your fork.

> Note: At this stage don't get disheartened if you hit errors and problems along the way. This often requires a lot of trial and error, as you learn about the particulars of your particular cluster. All are different, and all require TLC to get everything optimal!

To test your config, pick your pipeline of choice, and run the given pipeline's integrated mini-test profile but using your config. For example:

```bash
nextflow run nf-core/eager \
-profile <your_cluster_name>,test_tsv \
--custom_config_base 'https://raw.githubusercontent.com/<your_github_user>/configs/<your_branch>'
```

> If you also wish to test a profile from your `profiles{}` scope, include that in the `-profile` flag. For example:

Reviewer: ```suggestion
If you also wish to test a profile from your `profiles` scope, include that in the `-profile` flag. For example:
```

Now time to sit back and feel good for helping yourself and all institutional Nextflow users make their experience of running pipelines as smooth and efficient as possible 😎.

## Additional tips and tricks

### If your cluster often have too-full harddisk space issues

Some clusters often have very strict user HDD-space footprint restrictions (or just is often full). You can minimise the footprint that Nextflow runs use by using the `cleanup` directive.

This directive goes _outside_ any of the config scopes, and is simply defined as:

```nextflow
cleanup = true
```

This directive will, on a _successful_ completion of a Nextflow run, automatically delete all intermediate files stored in the `work/` directory. Note that none of the 'published' files in the `--outdir` results directory will be deleted if the pipeline specifies to _copy_ results files, _however_ removal of these intermediate files will mean debugging 'silent fails' more difficult.

To get around this, we suggest that you can optionally make an additional profile called debug:

```nextflow
debug {
  cleanup = false
}
```

which allows you to override the default `cleanup = true` behaviour, if you've set this as so.

> Alternatively, if your cluster utilises a scratch space to store intermediate files, this can be specified a the `process{}` scope of the profile with the [`scratch`](https://www.nextflow.io/docs/latest/process.html#scratch) directive. For example:

Reviewer: ```suggestion
Alternatively, if your cluster utilises a scratch space to store intermediate files, this can be specified a the `process` scope of the profile with the [`scratch`](https://www.nextflow.io/docs/latest/process.html#scratch) directive. For example:
```
